### PR TITLE
Added Physiofit 2.1.0

### PR DESCRIPTION
**Pull request by llegregam**
FOR CONTRIBUTOR:
* [ y] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ y] - License permits unrestricted use (educational + commercial)
* [ y] - This PR adds a new tool or tool collection
* [ n] - This PR updates an existing tool or tool collection
* [ n] - This PR does something else (explain below)

### CODE REVIEW

**Comment by llegregam** [comment]
@bgruening @yguitton  Please review :)

**Review by yguitton — CHANGES_REQUESTED** [review]
Hello @llegregam, in order to avoid code injection can you add simple quotes around variables sent to the script  like  --data '$input_selection.datafile' or -mc '$basic_settings.mc_select.iters'
thanks

File: tools/physiofit/physiofit.xml [hunk 1]
```diff
@@ -0,0 +1,202 @@
+<tool id='physiofit' name='PhysioFit' version='2.1.0+galaxy0' python_template_version='3.5'>
+    <requirements>
```
Comment (bgruening) [inline tools/physiofit/physiofit.xml:2]: "description" is missing

File: tools/physiofit/physiofit.xml [hunk 2]
```diff
@@ -0,0 +1,202 @@
+<tool id='physiofit' name='PhysioFit' version='2.1.0+galaxy0' python_template_version='3.5'>
```
Comment (bgruening) [inline tools/physiofit/physiofit.xml:1]: please add a profile to the tool tag

File: tools/physiofit/physiofit.xml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    ]]></command>
+    <inputs>
+
+        <conditional name='input_selection'>
+            <param name='input_selector' type='select' label='Choose the input data type to upload (tsv or json config file. For more information please check the documentation'>
+                <option value='tsv'>tsv file</option>
+                <option value='json'>config file (json)</option>
+            </param>
+            <when value='tsv'>
+                <param name='datafile' type='data' format='tsv, txt' multiple='true' label='Upload data file containing biomass and concentration values'/>
```
Comment (bgruening) [inline tools/physiofit/physiofit.xml:55]:
```suggestion
                <param name='datafile' type='data' format='tabular,txt' multiple='true' label='File containing biomass and concentration values'/>
```

File: tools/physiofit/physiofit.xml [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+            </when>
+            <when value='json'>
+                <param name='datafile' type='data' format='tsv, txt' multiple='true' label='Upload data file containing biomass and concentration values'/>
```
Comment (bgruening) [inline tools/physiofit/physiofit.xml:58]:
```suggestion
                <param name='datafile' type='data' format='tabular,txt' multiple='true' label='File containing biomass and concentration values'/>
```

File: tools/physiofit/physiofit.xml [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+            <conditional name='deg_select'>
+                <param name='deg' type='boolean' argument='--deg' label='Flag to add degradation constants' truevalue='true' falsevalue=''/>
+                <when value='true'>
+                    <param type='text' name='degconsts' value='{}' label='Enter degradation constants in dictionary format (see docs for more information)'/>
+                </when>
+            </conditional>
+
+            <conditional name='mc_select'>
+                <param name='montecarlo' type='boolean' argument='--montecarlo' label='Flag for sensitivity analysis' truevalue='true'/>
```
Comment (bgruening) [inline tools/physiofit/physiofit.xml:76]: please use a select parameter here not boolean: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#booleans

**Review by llegregam — COMMENTED** [review]
changes done

File: tools/physiofit/physiofit.xml [hunk 6]
```diff
@@ -52,29 +69,35 @@
...
+                <param name='datafile' type='data' format='tabular, txt' multiple='true' label='Upload data file containing biomass and concentration values'/>
                 <param name='cfgfile' type='data' format='json' label='Upload configuration file' />
             </when>
         </conditional>
 
 
         <section name='basic_settings' title='Basic Settings'>
 
-            <param name='lag' type='boolean' argument='--lag' label='Flag to estimate lag phase' truevalue='-l' falsevalue=''/>
+            <param name='lag' type='boolean' label='Flag to estimate lag phase'/>
```
Comment (bgruening) [inline tools/physiofit/physiofit.xml:83]: The previous version looks better. Using argument is also encouraged :-)

Comment (llegregam) [reply tools/physiofit/physiofit.xml]: True, I don't know why I changed that ^^

File: tools/physiofit/physiofit.xml [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+        </section>
+
+    </inputs>
+
+    <outputs>
+    <data name='plots' label='Plots' format='pdf'/>
+    <data name='fluxes' label='Flux_results' format='tabular'/>
+    <data name='stats' label='Stat_results' format='tabular'/>
+    <data name='config' label='Config_file' format='json'/>
```
Comment (lecorguille) [inline tools/physiofit/physiofit.xml:172]: Indentation, here

File: tools/physiofit/physiofit.xml [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        <citation type='bibtex'>
+    @misc{githubphysiofit,
+      author = {Le Gregam, Loic},
+      year = {2022},
+      title = {PhysioFit},
+      publisher = {Peiro et al.},
+      journal = {Appl Environ Microbiol},
+      url = {https://journals.asm.org/doi/10.1128/AEM.00768-19},
+    }</citation>
+        </citations>
```
Comment (lecorguille) [inline tools/physiofit/physiofit.xml:239]: Indentation here

**Review by lecorguille — CHANGES_REQUESTED** [review]

**Review by yguitton — APPROVED** [review]

**Review by yguitton — APPROVED** [review]

File: tools/physiofit/physiofit.xml [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                <param name='fm_select|fm' value='false'/>
+                <param name='cb_select|cb' value='false'/>
+                <param name='fb_select|fb' value='false'/>
+                <param name='verbose' value='false'/>
+            </section>
+            <output name='fluxes'>
+                <assert_contents>
+                    <has_n_columns n='7'/>
+                </assert_contents>
+            </output>
```
Comment (lecorguille) [inline tools/physiofit/physiofit.xml:195]: Your test isn't really strong, isn't it?

Why don't you use `test-data/flux_results.tsv`?

Comment (llegregam) [reply tools/physiofit/physiofit.xml]: I had trouble using the flux_results.tsv file as comparison, I do not know why. I added a file size check and a line number check, it should be good enough now right?

Comment (lecorguille) [reply tools/physiofit/physiofit.xml]: My guess is that the results have too many floats.
Values that are too precise may vary from machine to machine

Comment (Lain-inrae) [reply tools/physiofit/physiofit.xml]: In the case of floats imprecision, you can check the existence of some lines that matches regex:
https://docs.galaxyproject.org/en/latest/dev/schema.html#id141

for example:
```xml
                <assert_contents>
                    <has_text_matching expression="X_0\s0\.0198\d+\s0\.02\d+\s0\.0019\d+\s0\.02\d+\s0\.016\d+\s0\.02399\d+" />
                </assert_contents>
```
You can try to match each value (like I did), or you can match only some specific values - just to check you actually have values. You can even just replace whole numbers by regex to check there are no "NA" or "NULL" or missing values:
```X_0(\s\d+\.\d+){6}```

Comment (Lain-inrae) [reply tools/physiofit/physiofit.xml]: As the doc says: «  If the regular expression is expected to match a particular number of times, this value can be specified using n. »
```
                <assert_contents>
                    <has_text_matching expression="[a-zA-Z0-9_]+(\s\-?\d+(e\-?|\.)\d+){6}" n="6" />
                </assert_contents>
```
With this you can check you have six lines, each beginning with at least one or more alphanumeric character or underscores, followed by the pattern: a space character (tabs are considered as spaces characters), followed by a number,  six times.

**Pull request merged**